### PR TITLE
Resolving [most?] PHP 7 compatibility issues

### PR DESCRIPTION
**Pull request by gohanman**
Following up on #358 I think this covers most issues or at least everything php7cc finds. Includes:

* Replacing mysql with mysqli. Lifted directly from existing work in #171. Seems to work fine.
* Removing `&` from `$var &= new Foo()` instances. The assign by reference there is a syntax error in PHP7.
* Using `explode` in place of `split`.
* Replacing `ereg` functions. For instances that were validating email addresses I used `filter_var` instead. Other instances are changed to use `preg` functions.
* Get rid of `set_magic_quotes_runtime`. I conditionally disabled these calls based on PHP version although deleting them outright would probably be fine too. PHP versions that support magic quotes are all end-of-life'd already.

With this set of changes I have OpenCATS running on PHP 7.0.30

### CODE REVIEW

**Comment by RussH** [comment]
@mlespiau can you amend Travis CI to build this branch with PHP 7.x ?

**Comment by RussH** [comment]
Thanks @gohanman just need to get travis running php 7 checks.. and I should probably roll in the previous pending PR's first.. will be back in touch!

**Comment by RussH** [comment]
Looks good but will push a further release with other fixes included, then a new one that also includes php7 support

**Comment by shantadam** [comment]
@RussH why delay it further? It seems to me that it's a feature that many of the project supporters have been awaiting for a while.

File: modules/settings/SettingsUI.php [hunk 1]
```diff
@@ -2965,7 +2965,7 @@ private function wizard_addUser()
         $users = new Users($this->_siteID);
 
         /* If adding an e-mail username, verify it is a valid e-mail. */
-        if (strpos($loginName, '@') !== false && !eregi("^[_a-z0-9-]+(.[_a-z0-9-]+)*@[a-z0-9-]+(.[a-z0-9-]+)*(.[a-z]{2,4})$", $loginName))
+        if (strpos($loginName, '@') !== false && filter_var($username, FILTER_VALIDATE_EMAIL) === false)
```
Comment (BentCoder) [inline modules/settings/SettingsUI.php]:
```suggestion
        if (strpos($loginName, '@') !== false && filter_var($loginName, FILTER_VALIDATE_EMAIL) === false)
```

File: rebuild_old_docs.php [hunk 2]
```diff
@@ -50,12 +51,12 @@ function rebuild_old_docs() {
 
 
 //$con = mysql_connect("localhost","root","root");
-$con = mysql_connect(DATABASE_HOST, DATABASE_USER, DATABASE_PASS);
+$con = mysqli_connect(DATABASE_HOST, DATABASE_USER, DATABASE_PASS);
 if (!$con)
 {
-  die('Could not connect: ' . mysql_error());
+  die('Could not connect: ' . mysqli_error($co$conn));
```
Comment (BentCoder) [inline rebuild_old_docs.php]: `$co$conn` ?

**Comment by BentCoder** [comment]
I am not sure if `mysqli_select_db` has been used correctly in this PR. e.g. a `string` is being passed when `mysqli` is expected.

**Comment by gohanman** [comment]
It looks to me like it's wrong in a couple tests but correct in the non-test code which is why it seemed to work. Like I said in the opening post I didn't write the mysqli commits.

**Comment by BentCoder** [comment]
After applying the changes listed in this PR, the Hiring Overview graph stopped appearing. Any particular reason for it? Anybody knows why?

screen shot 2018-12-26 at 16 38 39

**Comment by BentCoder** [comment]
> After applying the changes listed in this PR, the Hiring Overview graph stopped appearing. Any particular reason for it? Anybody knows why?
> 
> screen shot 2018-12-26 at 16 38 39

Found it. Could someone please update `lib/artichow/inc/Font.class.php:252` file to replace `split` with `explode` please.

Thanks

**Comment by gohanman** [comment]
As far as I can tell it already is https://github.com/gohanman/OpenCATS/blob/full-7/lib/artichow/inc/Font.class.php#L252

**Comment by BentCoder** [comment]
Looks like my copy is old.

**Comment by RussH** [comment]
@BentCoder @gohanman have you stuck on PHP7.0, as I (finally) have the test php 7 branch running on PHP 7.2 and building in travis. There are some further issues from 7.0 -> 7.2, checking in to see if you've worked php 7.2 compatibility at-all? Otherwise please submit any corrective PR's to the test-php-7 branch!

**Comment by gohanman** [comment]
Sorry, I haven't done anything further.

**Comment by BentCoder** [comment]
For 7.2:

I had to make properties below `public`.

```php
class Contacts
{
    private $_db;
    private $_siteID;

    ...
}
```

I had to:

- Add `include_once('./lib/Contacts.php');`.
- Add a new `public $contacts;` class property.
- Add `$this->contacts = new Contacts($siteID);` to `__construct` - as last line.
- Convert every `Contacts::....` call to `$this->contacts->....`.

in `JobOrders` class.

**Comment by RussH** [comment]
If you have them in a repo - can you send a PR into the project? Also - if
you're not in the slack channel, it'd be good to have you in there, too!
just drop me an email and I'll invite you


----
Russ

**Comment by RussH** [comment]
Did you install mcrypt through PECL, or have you reworked
/lib/encryption.php?


----
Russ

**Pull request merged**